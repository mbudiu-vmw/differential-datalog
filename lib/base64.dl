extern function decode(s: string): Result<string, string>
extern function encode(s: string): string
