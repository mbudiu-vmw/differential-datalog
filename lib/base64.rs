use base64::DecodeError as b64DecodeError;
use base64::decode as b64decode;
use base64::encode as b64encode;

pub fn encode(s: &String) -> String {
    b64encode(s)
}

pub fn decode(s: &String) -> crate::ddlog_std::Result<String, String> {
    match b64decode(s) {
        Ok(r) => crate::ddlog_std::Result::Ok{ res: String::from_utf8(r).unwrap() },
        Err(e) => match (e) {
            b64DecodeError::InvalidByte(p, b) => crate::ddlog_std::Result::Err{
                err: format!("Invalid byte {} at position {}", b, p)
            },
            b64DecodeError::InvalidLength => crate::ddlog_std::Result::Err{
                err: format!("Invalid length")
            },
            b64DecodeError::InvalidLastSymbol(p, b) => crate::ddlog_std::Result::Err{
                err: format!("Invalid last byte {} at position {}", b, p)
            }
        }
    }
}
