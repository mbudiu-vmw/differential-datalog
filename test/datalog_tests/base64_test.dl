import base64

output relation Base64(str: string, encoding: string)
Base64("TEST", decode(encode("TEST")).unwrap_or("")).
Base64("TEST", encode("TEST")).
