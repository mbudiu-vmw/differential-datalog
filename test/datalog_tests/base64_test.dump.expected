base64_test::Base64{.str = "TEST", .encoding = "TEST"}
base64_test::Base64{.str = "TEST", .encoding = "VEVTVA=="}
